### ISSUE TITLE
shutdown-after-unsuccessful-join setting VS health checks 

### ISSUE THREAD

**Issue by ignasi35** [lightbend/openshift-deployment-guide#35]
The docs on forming a cluster instruct the user to set `shutdown-after-unsuccessful-join-seed-nodes = 60s`.

That value competes with Kubernetes' probing and Health Checks so both settings (`shutdown-after-unsuccessful-join-seed-nodes = 60s` and YAML's `readinessProbe.*` values) must be tuned carefully.

Actually, does it make sense to use `shutdown-after-unsuccessful-join-seed-nodes = 60s` when using Kubernetes `readinessProbe`?

**Comment by jroper** [comment]
Agreed. It defaults to `off`, we should leave it as the default.

**Comment by jroper** [comment]
Related is `lagom.cluster.exit-jvm-when-system-terminated`. In the sample app I originally based things off, this was set to `true`, but I left the config out, mainly as a convenience for the guide since the forming a cluster section was mostly non Lagom specific but that option is Lagom specific. I think we should let Kubernetes handle non graceful exits via liveness and readiness checks, and allow it to kill the process.

**Comment by TimMoore** [comment]
The readiness probe doesn't shut down non-responsive containers, it just avoids serving them traffic. So a node that doesn't shut down after unsuccessfully joining the cluster will just hang around as a useless replica until terminated.

The default liveness probe (which does kill the container if it fails) does not check for cluster membership. We have had bad experiences in the past with customers and other users when Lightbend Orchestration did check for this in the liveness probe. If the timeout was set too low, Kubernetes would often kill a node that was still bootstrapping, leading to an infinite cycle of starting a container and then killing it, with no cluster ever forming. I think it makes sense for a service to exit if it knows it will never become functional. Liveness checks are really intended for hung processes and other similar cases where the process is unable to manage its own lifecycle.

**Comment by TimMoore** [comment]
CC: @patriknw @chbatey @raboof what are your thoughts?

**Comment by chbatey** [comment]
We added it as a recommended setting after some edge cases were reported of a node discovering some nodes that then shutdown before it is able to join. You're then left with a node that will just sit there doing nothing, not part of a cluster.

Readiness probe won't help.
Liveness probe would work on start up but then we don't want a node to be killed after joining, we want SBR to make that decision. 

IMO leaving it set to 30s or 60s is the best option unless we change bootstrap to retry.

For context https://github.com/akka/akka-management/issues/356

**Comment by patriknw** [comment]
A guide that is using Akka Management should include example configuration that is enabling those:

```
akka.cluster.shutdown-after-unsuccessful-join-seed-nodes = 60s
akka.coordinated-shutdown.exit-jvm = on
```

**Comment by ignasi35** [comment]
Thanks all for the extra info. More details:

1. Play includes a [`reference-overrides.conf`](https://github.com/playframework/playframework/blob/2efaa68b6b4e0c5035f66654c4b1c6366f646821/core/play/src/main/resources/play/reference-overrides.conf#L59) that sets `akka.coordinated-shutdown.exit-jvm = off`. The reason is Play's JVM shutdown hooks take over the shutdown sequence and having Akka invoke `System.exit()` causes a deadlock. When the cause is a cluster node Downing then it's already handled.

2. There's no `override` for `shutdown-after-unsuccessful-join-seed-nodes` in either Play or Lagom.

3. Lagom's `exit-jvm-when-system-terminated` defaults to `off` but is recommeneded `on` in production.

Summing up, starting Lagom 1.5.0 I think production apps should run with:

```
lagom.cluster.exit-jvm-when-system-terminated = on
akka.cluster.shutdown-after-unsuccessful-join-seed-nodes = 60s
## This is not necessary, but if added it should be `off` or deadlocks may happen.
# akka.coordinated-shutdown.exit-jvm = on
```

Does that sound correct?

**Comment by renatocaval** [comment]
I added this to the Lagom instructions (https://github.com/lightbend/openshift-deployment-guide/pull/42/commits/f505f2a297314d5b2a592217816e3b9743a668ba) and I will send a PR to Lagom.

We can't add it to Lagom 1.5.0 anymore so we will need to keep it in your guides.

**Comment by renatocaval** [comment]
Here is the PR on Lagom: https://github.com/lagom/lagom/pull/1811

**Comment by renatocaval** [comment]
I will close this one. The guide covers it now and we have also added it to Lagom.
